### DISCUSSION TITLE
For typescript extras, how to add auto import on save?

### DISCUSSION THREAD

**Discussion by neonvoidx** [LazyVim/LazyVim#4103] [Q&A] [answered]
Wondering how I can make it so when i save a typescript file it auto imports any missing imports as well as auto removes any unused imports? Using typescript extras atm
From vtsls:
```
source.removeUnused.ts
source.addMissingImports.ts
```

**Comment by dotfrag** [comment, marked as answer]
This autocommand works, but I believe there is a better way to do it that I'm not aware of:

```lua
vim.api.nvim_create_autocmd("BufWritePre", {
  group = augroup("organise_imports"),
  pattern = "*.ts",
  callback = function()
    vim.lsp.buf.code_action({ apply = true, context = { only = { "source.addMissingImports.ts" }, diagnostics = {} } })
    vim.lsp.buf.code_action({ apply = true, context = { only = { "source.removeUnused.ts" }, diagnostics = {} } })
  end,
})
```

**Reply by neonvoidx** [reply]
nice ok, that works, ill mark as answer if no one else replies with the "correct" way shortly

**Reply by dotfrag** [reply]
No problem! I'd wait for a day or two.

**Reply by dpetka2001** [reply]
I don't know since I don't do Typescript, but from quickly searching trough past Discussions, I found #1943 which seems to suggest the same thing as suggested answer. So, I believe @dotfrag's answer is the way to go. Again, I'm not saying that there might not be a better answer, since I don't even know myself.

**Reply by artyultra** [reply]
im still so new at vim, where would I put this code? in my lazy.lua file?

**Reply by dpetka2001** [reply]
In `~/.config/nvim/lua/config/autocmds.lua`.

**Comment by alfieqashwa** [comment]
"leader + co" to Organize imports OR "leader + cu" to Remove unused imports

just install TypeScript Lang with :LazyExtras

source -> https://www.lazyvim.org/extras/lang/typescript
